feat: add openfga to compose

ryanfang5 opened the pull request.

igboyes commented on docker-compose.yml:
```diff
@@ -178,6 +178,24 @@ services:
+      OPENFGA_LISTOBJECTSMAXRESULTS: 10000
+      OPENFGA_LISTOBJECTSDEADLINE: "10s"
+      OPENFGA_LOG_FORMAT: "json"
+    image: openfga/openfga:latest
```
Please pin to the latest version number. We don't want any surprises when `latest` starts meaning something else.

igboyes commented on docker-compose.yml:
```diff
@@ -178,6 +178,24 @@ services:
+      OPENFGA_LOG_FORMAT: "json"
+    image: openfga/openfga:latest
+    container_name: openfga
+    command: run --datastore-engine postgres --datastore-uri 'postgres://virtool:virtool@postgres:5432/postgres?sslmode=disable'
```
Is this sticking the data in the `virtool` database along with our application data? I think the OFGA and VT data should be in separate DBs in the same Postgres server.

ReeceHoffmann replied:
Not strictly required in dev, but I would move the `--datastore-engine` and `--datastore-uri` commands to be passed as env variables. Presumably in prod the secure DB password is going to be provided via env variable or similar.

https://openfga.dev/docs/getting-started/setup-openfga

ReeceHoffmann commented on docker-compose.yml:
```diff
@@ -178,6 +178,24 @@ services:
+      - default
+    ports:
+      - "8080:8080"
+      - "3000:3000"
```
I think it makes sense to expose the 3000 playground port in dev, so we should leave this as is, but just as a note, the docs recommend against exposing 3000 in prod

ReeceHoffmann commented on docker-compose.yml:
```diff
@@ -178,6 +178,24 @@ services:
+      - all
+      - test
+      - backend
+      - ui
```
We should also add a depends_on field to ensure this doesn't try to start prior to PostgreSQL starting. Something like what we have for the server: 

```
depends_on:
      - postgres
```

ReeceHoffmann commented on docker-compose.yml:
```diff
@@ -178,6 +178,24 @@ services:
+      OPENFGA_LISTOBJECTSDEADLINE: "10s"
+      OPENFGA_LOG_FORMAT: "json"
+    image: openfga/openfga:latest
+    container_name: openfga
```
Do we need to specify `container_name`? Wouldn't it generate something similar using the name provided on line 181?

ReeceHoffmann commented on docker-compose.yml:
```diff
@@ -178,6 +178,24 @@ services:
+    image: openfga/openfga:latest
+    container_name: openfga
+    command: run --datastore-engine postgres --datastore-uri 'postgres://virtool:virtool@postgres:5432/postgres?sslmode=disable'
+    networks:
+      - default
```
Do we need to specify the default networks explicitly?

ReeceHoffmann reviewed (changes requested):
Mostly looks good! I also think we should probably also add the migrate service to the compose file. Looks to me like the migrate command has to be run at least once on a given postgres server before OFGA can interact with it. Might also be important for OFGA updates, but docs are not clear on that.

https://openfga.dev/docs/getting-started/setup-openfga

ryanfang5 commented:
Took way too long due to not attaching persisting volume to the postgres-openfga container. Also moved commands into environment variables and removed default networks. Let me know if any naming should be changed.
